### PR TITLE
PAYARA-4120 Enabled default ErrorService until configured in other way

### PR DESCRIPTION
**Pull request by dmatej**
The main difference in behavior is this:

**Original**

1. Invoke creation of a ServiceLocator (usually via some ModulesRegistry implementation)
2. If there is a problem, you will not know about it, you will receive a ServiceLocator which may or may not work. Trying to access DynamicConfigurationService may endup with NPE/ISE as in https://github.com/eclipse-ee4j/glassfish-hk2/issues/464 but you will not get real cause.
3. If there is not a problem, you will receive a ServiceLocator instance. You can reconfigure it or use it, whatever.

**Now**

1. Invoke creation of a ServiceLocator (usually via some ModulesRegistry implementation)
2. If there is a problem, it will cause an exception. Deal with it.
3. If there is not a problem, you will receive a ServiceLocator instance. You can reconfigure it or use it, whatever.

- originally ServiceLocator initialization ignored all errors, so when the DynamicConfigurationService  was not accessible because of classloading issues (on JDK11), it ended up with NPE without any usable information about the cause (not even in configured logs).

- now the original behavior can be reached by configuring the IgnoringErrorService, but creation of ServiceLocator would throw exception if it could not be instantiated.

- initialization is sucessful and the service is configured OR you catch an exception, but you should not get unconfigured failing locator.

### CODE REVIEW

**Review by pdudits — APPROVED** [review]

**Review by Cousjava — CHANGES_REQUESTED** [review]
Fails build tests

**Review by AlanRoth — APPROVED** [review]
Code LGTM - But I get the same failures as Jonathan when building
```
[ERROR] Failures:
[ERROR]   NegativeFactoryTest.testFactoryThatThrowsInPerLookupScope:80
[ERROR]   NegativeFactoryTest.testFactoryThatThrowsInPerThreadScope:96
[ERROR]   NegativeFactoryTest.testFactoryWithBadName:50 A MultiException has 2 exceptions.  They are:
1. java.lang.IllegalArgumentException: The scope name given in the descriptor (org.glassfish.hk2.api.PerLookup) did not match the scope annotation on the class (javax.inject.Singleton) in class BadlyNamedFactory
2. java.lang.IllegalArgumentException: Errors were discovered while reifying SystemDescriptor(
        implementation=org.glassfish.hk2.tests.locator.negative.factory.BadlyNamedFactory
        contracts={org.glassfish.hk2.api.Factory,org.glassfish.hk2.tests.locator.negative.factory.BadlyNamedFactory}
        scope=org.glassfish.hk2.api.PerLookup
        qualifiers={}
        descriptorType=CLASS
        descriptorVisibility=NORMAL
        metadata=
        rank=0
        loader=null
        proxiable=null
        proxyForSameScope=null
        analysisName=null
        id=6
        locatorId=266
        identityHashCode=807239968
        reified=false)

[ERROR] Errors:
[ERROR]   NegativeClassAnalysisTest.testBadConstructorNull ▒ NoClassDefFound Could not i...
[ERROR]   NegativeClassAnalysisTest.testBadConstructorThrow ▒ NoClassDefFound Could not ...
[ERROR]   NegativeClassAnalysisTest.testBadFieldNull ▒ NoClassDefFound Could not initial...
[ERROR]   NegativeClassAnalysisTest.testBadFieldThrow ▒ NoClassDefFound Could not initia...
[ERROR]   NegativeClassAnalysisTest.testBadMethodNull ▒ NoClassDefFound Could not initia...
[ERROR]   NegativeClassAnalysisTest.testBadMethodThrow ▒ ExceptionInInitializer
[ERROR]   NegativeClassAnalysisTest.testBadPCThrow ▒ NoClassDefFound Could not initializ...
[ERROR]   NegativeClassAnalysisTest.testBadPDThrow ▒ NoClassDefFound Could not initializ...
[ERROR]   NegativeClassAnalysisTest.testSelfAnalyzer ▒ NoClassDefFound Could not initial...
[ERROR]   ServiceLocatorUtilitiesTest.testErrorRethrower:575 ▒ Multi A MultiException ha...
[ERROR]   ServiceLocatorUtilitiesTest.testErrorRethrowerViaModule:618 ▒ Multi A MultiExc...
[INFO]
[ERROR] Tests run: 517, Failures: 3, Errors: 11, Skipped: 4
```

**Comment by dmatej** [comment]
Funny, it seems that some tests were always broken, but because ServiceLocator tolerated errors in it's initialization, those errors were not reported, so those tests could test another problems.
I will fix them and update the PR (I hope I will not find any other real problem).

Side note: tests depend on artifacts in local repository, so the build sometimes needs to be run twice. I could fix it too, but in another PR much later ...

**Comment by dmatej** [comment]
Tested with JDK11 and Payara commit 6c1cc81aed built with -Dhk2.version=2.6.1.payara-p1 (later commits have fixed dependencies).

With this fix it prints beautiful stacktrace with messages about missing PostConstruct dependency which is requested behavior.
Without that it printed NPE, because "it was secure".

StaticModulesRegistry configures the IgnoringErrorService to reach original behavior, but later after the ServiceLocator knows it can reach DynamicConfigurationService. This error service at least can log the error information if configured so, so it still fullfils both the request of security and the request to reach the cause of problem.

**Review by Cousjava — CHANGES_REQUESTED** [review]
I do not think this is a good solution. HK2 already threw the error, the issue is Payara swallowing it not HK2. It would be better to add the error service when HK2 is started within Payara.

**Comment by dmatej** [comment]
Payara is not swallowing it, it does HK2. It could not instantiate the DynamicCfgService, because of unreachable dependencies, but it swallowed all errors, so finally it created unusable ServiceLocator instance throwing nullpointers. 
Payara could not do anything with this and we spent days or weeks until we have found the real cause (asadmin classpath).

**Comment by dmatej** [comment]
About security: providing broken service locator which internally broke several own rules but ignored them is security issue from my point of view. They even had tests for that, but they were targeting behavior of provided locator, not it's initialization.

So when I did this, tests failed, because this prevented to provide locator with errors.

Current PR tests both scenarios, with ignoring error service or rethrow error service. Payara can provide another implementation but will always receive correct locator to configure.

**Comment by dmatej** [comment]
Jenkins test please

> Fails build tests

Last push fixed this.

File: hk2-locator/src/test/java/org/glassfish/hk2/tests/locator/negative/factory/BadlyNamedFactory.java [hunk 1]
```diff
@@ -20,21 +20,22 @@
...
 /**
  * This factory has a provide method with a {@link Named}
  * annotation with no value, which is illegal
- * 
+ *
  * @author jwells
  *
  */
-@Singleton
+@PerLookup
```
Comment (Cousjava) [inline hk2-locator/src/test/java/org/glassfish/hk2/tests/locator/negative/factory/BadlyNamedFactory.java:32]: Why is this changed?

File: hk2-api/src/main/java/org/glassfish/hk2/utilities/IgnoringErrorService.java [hunk 2]
```diff
@@ -0,0 +1,67 @@
+/*
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS HEADER.
+ *
+ *    Copyright (c) 2019 Payara Foundation and/or its affiliates. All rights reserved.
+ *
+ *     The contents of this file are subject to the terms of either the GNU
```
Comment (Cousjava) [inline hk2-api/src/main/java/org/glassfish/hk2/utilities/IgnoringErrorService.java]: This should be EPL for 2.6.1 branch

File: hk2-api/src/main/java/org/glassfish/hk2/utilities/RethrowErrorService.java [hunk 3]
```diff
@@ -16,43 +16,41 @@
...
 public class RethrowErrorService implements ErrorService {
 
-    /* (non-Javadoc)
-     * @see org.glassfish.hk2.api.ErrorService#onFailure(org.glassfish.hk2.api.ErrorInformation)
-     */
     @Override
-    public void onFailure(ErrorInformation errorInformation)
-            throws MultiException {
+    public void onFailure(ErrorInformation errorInformation) throws MultiException {
+        Logger.getLogger().debug("onFailure: " + errorInformation);
```
Comment (Cousjava) [inline hk2-api/src/main/java/org/glassfish/hk2/utilities/RethrowErrorService.java]: The point of the rethrow error service is not to print out the error, but to rethrow it so it can be printed elsewhere, could you remove this line please.

File: hk2-locator/src/test/java/org/glassfish/hk2/tests/locator/negative/classanalysis/ConfigurablyBadClassAnalyzer.java [hunk 4]
```diff
@@ -35,32 +35,32 @@
...
     private boolean throwFromFields = false;
     private boolean throwFromPostConstruct = false;
     private boolean throwFromPreDestroy = false;
-    
+
     private boolean nullFromConstructor = false;
     private boolean nullFromMethods = false;
     private boolean nullFromFields = false;
-    
-    @Inject @Named(ClassAnalyzer.DEFAULT_IMPLEMENTATION_NAME)
```
Comment (Cousjava) [inline hk2-locator/src/test/java/org/glassfish/hk2/tests/locator/negative/classanalysis/ConfigurablyBadClassAnalyzer.java:49]: Why was this changed?

File: hk2-locator/src/test/java/org/glassfish/hk2/tests/locator/negative/classanalysis/NegativeClassAnalysisModule.java [hunk 5]
```diff
@@ -23,22 +23,23 @@
...
+
     @Override
     public void configure(DynamicConfiguration config) {
         config.addActiveDescriptor(ConfigurablyBadClassAnalyzer.class);
-        
+
         config.bind(BuilderHelper.link(SelfAnalyzer.class.getName()).
                 to(ClassAnalyzer.class.getName()).
-                analyzeWith(NegativeClassAnalysisTest.SELF_ANALYZER).
+                analyzeWith(this.analyzer).
```
Comment (Cousjava) [inline hk2-locator/src/test/java/org/glassfish/hk2/tests/locator/negative/classanalysis/NegativeClassAnalysisModule.java:41]: Why was this changed?

**Review by Cousjava — CHANGES_REQUESTED** [review]

File: hk2-locator/src/test/java/org/glassfish/hk2/tests/locator/negative/factory/BadlyNamedFactory.java (hunk 1 above)
Comment (dmatej) [reply hk2-locator/src/test/java/org/glassfish/hk2/tests/locator/negative/factory/BadlyNamedFactory.java:32]: It is not allowed annotation here. But as exceptions were swallowed, it was ignored. This annotation was not tested, the missing name atttribute was.

File: hk2-api/src/main/java/org/glassfish/hk2/utilities/RethrowErrorService.java (hunk 3 above)
Comment (dmatej) [reply hk2-api/src/main/java/org/glassfish/hk2/utilities/RethrowErrorService.java]: It is not redundant, because:

1. It sends log message to logger with error details, not stacktrace, it is useful for debugging.
2. Thrown exception is missing these information.

The only bad thing is that it invokes toString even if it is not needed. And HK2 logger does not support suppliers, so ok, I will remove it.

File: hk2-locator/src/test/java/org/glassfish/hk2/tests/locator/negative/classanalysis/ConfigurablyBadClassAnalyzer.java (hunk 4 above)
Comment (dmatej) [reply hk2-locator/src/test/java/org/glassfish/hk2/tests/locator/negative/classanalysis/ConfigurablyBadClassAnalyzer.java:49]: I will put it back, it seems I removed and now it selects the best candidate, which is the same.

File: hk2-locator/src/test/java/org/glassfish/hk2/tests/locator/negative/classanalysis/NegativeClassAnalysisModule.java (hunk 5 above)
Comment (dmatej) [reply hk2-locator/src/test/java/org/glassfish/hk2/tests/locator/negative/classanalysis/NegativeClassAnalysisModule.java:41]: Good question ... I had to think about it again.

Previous version: used (SELF_ANALYZER=) Narcis analyzer to analyze Narcis service (forbidden but exception was ignored). Then locator was used with it's parent in test.
The last test was targeting forbidden cyclic usage of the analyzer to reify itself. 

Current version: uses default analyzer to analyze Narcis service without ignoring exceptions. Then all tests except final two use "BadConfigurableAnalyzer" to test several error scenarios. The final two tests (one was added) target forbidden cyclic dependencies - the longer test in effect tests the original scenario with ignorance and disabling it. 
The shorter test is added and tests fast fail if there is cyclic dependency at the time of ServiceLocator creation.

**Comment by dmatej** [comment]
I'm bad in explaining things, sorry :crying_cat_face: ... I will update upper text again ...

What I would like to know is ...

* what other (except classloading and cyclic analyzers) issues can cause failure when creating ServiceLocator (and the failure would be ignored in previous version and broken locator provided)
* how to better handle errors in HK2; how should it be configured for Payara?

Note that any configured ErrorService replaces the original default RethrowErrorService, which is not registered as a service, it is only a default instance in errorHandlers.
I love and hate this fact at the same time :D

And the second thing - logging errors is safe, providing stacktraces to server clients is not. So maybe Payara should configure it's own ErrorService implementation.

**Review by Cousjava — APPROVED** [review]

**Comment by dmatej** [comment]
Already merged in Glassfish HK2 repository:
https://github.com/eclipse-ee4j/glassfish-hk2/pull/465
https://github.com/eclipse-ee4j/glassfish-hk2/pull/476

**Comment by dmatej** [comment]
Oh, you merged it sooner than I created another comment. Beware of merging pull requests into the same already released version id. If you would build this branch again, it will get into .m2 cache and will be used by the build server instead of already released version with the same id.
That's why it should be reversioned to a snapshot immediately after release.
(The good thing is that in current Payara nobody should notice any difference, because the original problem, that led to this change, was already fixed :) )
